### ISSUE TITLE
AttributeError on salt-master after upgrading from "2016.3.2" to "2016.3.3"

### ISSUE THREAD

**Issue by HWDexperte** [saltstack/salt#36461]
### Description of Issue
After performing an `apt-get dist-upgrade` on my Debian 8 salt master, the installation via apt failed with this trace:

```bash
# apt-get dist-upgrade
Unpacking salt-cloud (2016.3.3+ds-1) over (2016.3.2+ds-1) ...
Unpacking salt-master (2016.3.3+ds-1) over (2016.3.2+ds-1) ...
Unpacking salt-common (2016.3.3+ds-1) over (2016.3.2+ds-1) ...
[...]
Setting up salt-common (2016.3.3+ds-1) ...
Setting up salt-cloud (2016.3.3+ds-1) ...
Setting up salt-master (2016.3.3+ds-1) ...

Job for salt-master.service failed. See 'systemctl status salt-master.service' and 'journalctl -xn' for details.
invoke-rc.d: initscript salt-master, action "start" failed.
dpkg: error processing package salt-master (--configure):
 subprocess installed post-installation script returned error exit status 1
dpkg: dependency problems prevent configuration of salt-api:
 salt-api depends on salt-master; however:
  Package salt-master is not configured yet.

dpkg: error processing package salt-api (--configure):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 salt-master
 salt-api
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

and `systemctl status salt-master.service` gives me the same output as `cat /var/log/salt/master`:

```
2016-09-21 11:39:14,969 [salt.log.setup   ][ERROR   ][596] An un-handled exception was caught by salt's global exception handler:
AttributeError: 'NoneType' object has no attribute 'lower'
Traceback (most recent call last):
  File "/usr/bin/salt-api", line 8, in <module>
    salt_api()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 464, in salt_api
    sapi.start()
  File "/usr/lib/python2.7/dist-packages/salt/cli/api.py", line 79, in start
    super(SaltAPI, self).start()
  File "/usr/lib/python2.7/dist-packages/salt/utils/parsers.py", line 941, in start
    self.prepare()
  File "/usr/lib/python2.7/dist-packages/salt/cli/api.py", line 53, in prepare
    verify_log(self.config)
  File "/usr/lib/python2.7/dist-packages/salt/utils/verify.py", line 523, in verify_log
    level = LOG_LEVELS.get(opts.get('log_level').lower(), logging.NOTSET)
AttributeError: 'NoneType' object has no attribute 'lower'
```
### Setup
In my `/etc/apt/sources.list` file there is this entry:
```
deb http://repo.saltstack.com/apt/debian/latest jessie main
```

### Steps to Reproduce Issue
Include the mentioned line in `/etc/apt/sources.list` in your own file and execute an `apt-get dist-upgrade` from salt-masver version 2016.3.2.

### Versions Report
```
Salt Version:
           Salt: 2016.3.3
 
Dependency Versions:
           cffi: 0.8.6
       cherrypy: 3.5.0
       dateutil: 2.2
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.3
        libgit2: 0.24.0
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: 0.24.0
         Python: 2.7.9 (default, Jun 29 2016, 13:08:31)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
 
System Versions:
           dist: debian 8.6 
        machine: x86_64
        release: 3.16.0-4-amd64
         system: Linux
        version: debian 8.6
```

**Comment by dahrens** [comment]
It seems that the commit 2fc3a55338b6f49e6ddd5d3da11273b1713b84bf introduces the line that breaks in the described setup. Removing the `.lower()` call makes the service running again.

**Comment by dahrens** [comment]
Providing a proper `log_level` in salt master configuration also solves the issue with the current version of salt.

**Comment by cachedout** [comment]
Hrm, so this is happening on any upgrade where `log_level` isn't explicitly set?!

**Comment by HWDexperte** [comment]
@cachedout yep

**Comment by cachedout** [comment]
@HWDexperte Ouch! Well, that's not great.

@meggiebot @Ch3LL Did this upgrade path not get tested? Can we figure out what happened here?

**Comment by dahrens** [comment]
the linked commit above includes tests, but the case that there is nothing written in the configuration is not  part of the test.

**Comment by cachedout** [comment]
@dahrens Aye, I noticed that as well. We'll definitely want to improve those tests so we catch conditions like this.

**Comment by dahrens** [comment]
@cachedout makes sense - @HWDexperte is currently diving into the unittests and tries provide a test for this case, too.

**Comment by cachedout** [comment]
@dahrens That's great news! Thanks to both you and @HWDexperte .

**Comment by HWDexperte** [comment]
@cachedout a single test for this case was added to the pullrequest.

**Comment by cachedout** [comment]
@HWDexperte Thank you very much! I'm very grateful.

**Comment by Ch3LL** [comment]
To clarify does your config file actually have: `log_level: ''`  or are you just not even setting log_level? Trying to figure out why we would not have seen this during package testing.

**Comment by HWDexperte** [comment]
@Ch3LL both. First we had the entry `log_level: null` (which isn't even valid), then i tested it by removing this entry and it worked neither. With `log_level: info` it worked fine (without the mentioned fix as pull request here)

**Comment by Ch3LL** [comment]
I can only replicate this with `log_level: null`. If doing an upgrade without touching log_level it works just fine, which explains why we wouldn't have seen it during package testing. So it looks like there might be a  bigger issue as to how did `log_level` even get set to null? I'm assuming you didn't set it to null? Did you set these VMs up originally with salt-cloud by chance?

**Comment by dahrens** [comment]
@Ch3LL our salt master is about two years old and we installed it manually. And yes, we didn't set `log_level: null` on purpose as far as i know.

The reason for this weird value might be our initial version of salt or which is my guess an installation of [tinyclues/saltpad v1](https://github.com/tinyclues/saltpad/tree/saltpad_v1) which we had running 	approximately one and a half year ago. It had a feature where is writes the whole master configuration to provide predefined job templates, which we used at least once.

**Comment by Ch3LL** [comment]
Thanks for the additional information @dahrens so it could have been a particular version that set that value. You don't happen to know the initial version? I'm guessing you wouldn't know which is fine, and it seems we have resolved the issue now because I know current versions of salt do not set this value to null. I know we have had issues with salt-cloud setting values in the master config so it is not related to that. 

Well I think for now since we can't exactly track down why that was showing up and its not on current versions of salt does everyone think this has been resolved on their end in terms of resolving the issue with how we handle the log_level value in the master config? Are we okay to close this since its been fixed?
